ci/ci/ci.py

```diff
@@ -206,6 +206,31 @@
 gh_router = gh_routing.Router()
 
 
+@routes.post('/invalidate_and_retest')
+@check_csrf_token
+@web_authenticated_developers_only
+async def invalidate_and_retest(request):
+    app = request.app
+    dbpool = app['dbpool']
+    batch_client = app['batch_client']
+    post = await request.post()
+    branch_id = post['watched_branch'].strip()
+    pr_number = post['pull_request_number'].strip()
+    found = False
+    for wb in watched_branches:
+        if wb.short_str() == branch_id:
+            pr = wb.prs.get(pr_number)
+            if pr is not None:
+                found = True
+                pr.batch = None
+                pr._heal(app, batch_client, dbpool, on_deck=False)
```

Comment:
This isn't going to work.  It might on the common path, but it isn't going to be reliable.  This is maintaining state in the CI process, and you can't do that.  CI's job is to reconcile the state in Github with the state in batch.  So you have two options: you need to modify the batch you want to ignore (e.g. change an attribute so it is ignored), keep a list of batches to ignore in the database (I added a ci database to store authorized SHAs).

For the record, I'm against this as a permanent feature.  The right thing here is to fix the batch and pipeline tests to be reliable.

Reply:
> maintaining state in the CI process

Bah! not again. Hmm...

> For the record, I'm against this as a permanent feature. The right thing here is to fix the batch and pipeline tests to be reliable

Sure, but nobody seems eager to do that at the moment.

Reply:
can I just delete the batch?

Reply:
or cancel?